### PR TITLE
Assign each actor an owner that manages actor restart

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This is the first step towards moving more system state into the workers and is necessary to eventually implement actor supervision.

In this PR, each worker tracks the actors that it has created so far (its "children"), saving their actor creation task spec and the number of times the actor has been created before. If an actor fails and it has max_reconstructions > 0 set in its `ray.remote` decorator, then the actor's owner will restart the actor by resubmitting its actor creation task.

Actor failures are detected slightly differently now:
- process failure: the local raylet adds a `RECONSTRUCTING` or `DEAD` entry to the actor table. This is the same as before.
- node failure: the actor's owner adds a `RECONSTRUCTING` or `DEAD` entry to the actor table. This is different from before, when all nodes would try to add the entry.

This also refactors the `DirectActorClient` to move the RPC client and actor state into the `ActorHandle` owned by the `CoreWorker` instead.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/core_worker/actor_handle.cc
Comment (ericl) [inline src/ray/core_worker/actor_handle.cc]: unique_ptr isn't thread safe, so we need the Mutex here right? Even if the update is atomic, the write may not be visible to other threads.

File: src/ray/core_worker/actor_handle.h
Comment (ericl) [inline src/ray/core_worker/actor_handle.h]: Instead of storing ActorTableData can we store the relevant fields instead? Basically it would nice if someone reading this code didn't need to know what a GCS is.

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: IMO this warning isn't needed.

Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: When is this warning useful?

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: if (!actor_handle->IsAlive())

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: Can we log loudly when-ever we restart an actor (ERROR level seems suitable). This should be a highly visible event to the user.

File: src/ray/core_worker/transport/direct_actor_transport.h
Comment (ericl) [inline src/ray/core_worker/transport/direct_actor_transport.h]: EXCLUSIVE_LOCKS_REQUIRED(mu_)

File: src/ray/core_worker/transport/direct_actor_transport.h
Comment (ericl) [inline src/ray/core_worker/transport/direct_actor_transport.h]: TODO(swang)

**Review by ericl — COMMENTED** [review]
Seems fine. What is the testing plan?

File: src/ray/core_worker/transport/direct_actor_transport.h
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.h]: https://github.com/abseil/abseil-cpp/blob/master/absl/base/thread_annotations.h#L112

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (ericl) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: This is super weird to be passing in null clients. Instead of the current breakdown of SubmitTask(option<client>), SendPendingTasks(), and FailPendingTasks(), how about:

1. Keep the signature of SubmitTask the same as before.
2. Keep the internal pointer to the current RPC client.
3. Add a OnActorLocationUpdate(client) method to this class.

That way, the logic of this class becomes a lot more self contained, and core_worker.cc just needs to post the location update to the transport when it changes.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (ericl) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: So it's not pending tasks more like fail in flight tasks right?

Actually do we even need this? The RPC requests will fail by themselves right?

Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: I guess the old code did this already, maybe it was just to fail them faster. @zhijunfu ?

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: Can this be in a helper function?

File: src/ray/core_worker/actor_handle.h
Comment (ericl) [inline src/ray/core_worker/actor_handle.h]: Can this have a better name?

File: src/ray/core_worker/actor_handle.cc
Comment (ericl) [inline src/ray/core_worker/actor_handle.cc]: Why not always reset?

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [reply src/ray/core_worker/core_worker.cc]: Actually it would be great if this were pulled out into a class like ActorManager. It would also then become possible to unit test handling of actor restarts.

Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc]: Which part specifically did you want to pull out?

File: src/ray/core_worker/actor_handle.cc
Comment (stephanie-wang) [reply src/ray/core_worker/actor_handle.cc]:
```
        // We have to reset the actor handle since the next instance of the
        // actor will not have the last sequence number that we sent.
        // TODO: Remove the flag for direct calls. We do not reset for the
        // raylet codepath because it tries to replay all tasks since the last
        // actor checkpoint.
```

**Comment by stephanie-wang** [comment]
> Seems fine. What is the testing plan?

I think just Python tests right now? There is a very simple test for actor restart in `core_worker_test.cc`, too.

File: src/ray/core_worker/core_worker.cc
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc]: We can't just check `IsAlive` because that might mean the actor is restarting.

File: src/ray/core_worker/transport/direct_actor_transport.h
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.h]: I think this was only true before. I changed this function to hold a lock.

File: src/ray/core_worker/actor_handle.cc
Comment (ericl) [reply src/ray/core_worker/actor_handle.cc]: Ah, the mythical task replay.

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: ERROR?

Comment (ericl) [reply src/ray/core_worker/core_worker.cc]: Can we also log error for successful restarts? It should be a notable condition even if we recover from it, and ERROR is the only level that shows up on the console.

**Review by ericl — APPROVED** [review]
Looks great!

File: src/ray/core_worker/actor_handle.cc
Comment (stephanie-wang) [reply src/ray/core_worker/actor_handle.cc]: Oh oops, this was a debugging thing, I'll remove it.

I agree, but we'll need to do something different for these sorts of errors. There is no client code that can catch this error because it's called in an async handler.

File: src/ray/core_worker/actor_manager.cc
Comment (stephanie-wang) [reply src/ray/core_worker/actor_manager.cc]: It can.

File: src/ray/core_worker/actor_manager.cc
Comment (stephanie-wang) [reply src/ray/core_worker/actor_manager.cc]: Ask @ericl

File: src/ray/core_worker/actor_manager.cc
Comment (stephanie-wang) [reply src/ray/core_worker/actor_manager.cc]: I'll add a comment that MarkDead is idempotent.

File: src/ray/core_worker/test/core_worker_test.cc
Comment (stephanie-wang) [reply src/ray/core_worker/test/core_worker_test.cc]: The previous test was wrong and I believe it only worked because it was slow. It waits for a RECONSTRUCTING state, then an ALIVE state, but the actor can switch through these states correctly before the next time the condition is checked. Then, the first check for RECONSTRUCTING will hang.

Comment (stephanie-wang) [reply src/ray/core_worker/test/core_worker_test.cc]: Sorry that it's hacky, but I would have preferred to not have this test at all (we should be testing these things from python, not here). Just trying to fix it up enough that it passes right now.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc]: We're going to remove this code soon.

File: src/ray/core_worker/actor_manager.h
Comment (zhijunfu) [inline src/ray/core_worker/actor_manager.h]: maybe change `terminal` to `is_reconstructable`?

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (zhijunfu) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: @ericl I remember the issue was that it's not guaranteed that GRPC callback will be invoked on failure case, so we also leverage the GCS actor notification to treat tasks as failed.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (zhijunfu) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: I remember in existing logic we will buffer the pending requests when actor is reconstructing, and only treat these tasks as failed when actor is dead (and cannot be reconstructed). But the new logic might be ok too.

**Comment by zhijunfu** [comment]
I didn't have chance to look at most of the implementations yet. A quick question:

Do we assume the actor's owner to be alive in the node failure case? e.g. If the actor's owner resides on the same node with the actor. Does the actor owner saves the actor info that it created to some persistent place? e.g. GCS.  

> 
> * process failure: the local raylet adds a `RECONSTRUCTING` or `DEAD` entry to the actor table. This is the same as before.
> * node failure: the actor's owner adds a `RECONSTRUCTING` or `DEAD` entry to the actor table. This is different from before, when all nodes would try to add the entry.
>

**Comment by stephanie-wang** [comment]
> I didn't have chance to look at most of the implementations yet. A quick question:
> 
> Do we assume the actor's owner to be alive in the node failure case? e.g. If the actor's owner resides on the same node with the actor. Does the actor owner saves the actor info that it created to some persistent place? e.g. GCS.

Right now, yes, but eventually it should not matter. The design is such that if you know the actor's owner, you should be able to determine whether the actor is alive or not. The actor info is still saved in the GCS.

> 
> > * process failure: the local raylet adds a `RECONSTRUCTING` or `DEAD` entry to the actor table. This is the same as before.
> > * node failure: the actor's owner adds a `RECONSTRUCTING` or `DEAD` entry to the actor table. This is different from before, when all nodes would try to add the entry.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: Hmm yeah I recall now that I had to change this once we set the task counter according to the counter in the handle, instead of in the direct actor transport. We can think about changing it back later, but I don't think it's a big deal. From the user's perspective, failing all pending tasks might actually be better.

**Comment by zhijunfu** [comment]
> > I didn't have chance to look at most of the implementations yet. A quick question:
> > Do we assume the actor's owner to be alive in the node failure case? e.g. If the actor's owner resides on the same node with the actor. Does the actor owner saves the actor info that it created to some persistent place? e.g. GCS.
> 
> Right now, yes, but eventually it should not matter. The design is such that if you know the actor's owner, you should be able to determine whether the actor is alive or not. The actor info is still saved in the GCS.

In this case should we preserve the current logic for handling node failure? Otherwise if the actor's owner and the actor resides on the same node, nobody will be reconstructing the specific actor when the node fails because its owner would die as well. We can remove the logic after this issue is resolved later. What do you think?

**Comment by stephanie-wang** [comment]
> > > I didn't have chance to look at most of the implementations yet. A quick question:
> > > Do we assume the actor's owner to be alive in the node failure case? e.g. If the actor's owner resides on the same node with the actor. Does the actor owner saves the actor info that it created to some persistent place? e.g. GCS.
> > 
> > 
> > Right now, yes, but eventually it should not matter. The design is such that if you know the actor's owner, you should be able to determine whether the actor is alive or not. The actor info is still saved in the GCS.
> 
> In this case should we preserve the current logic for handling node failure? Otherwise if the actor's owner and the actor resides on the same node, nobody will be reconstructing the specific actor when the node fails because its owner would die as well. We can remove the logic after this issue is resolved later. What do you think?

I don't think we should block this PR just to resolve that issue. The way to handle that particular issue is for the owner's owner to restart it. Once the owner is restarted, it should also restart the child actor.

If there are more questions about the actor supervision approach, can you please open a github issue for it? Thanks!

**Comment by stephanie-wang** [comment]
> > > I didn't have chance to look at most of the implementations yet. A quick question:
> > > Do we assume the actor's owner to be alive in the node failure case? e.g. If the actor's owner resides on the same node with the actor. Does the actor owner saves the actor info that it created to some persistent place? e.g. GCS.
> > 
> > 
> > Right now, yes, but eventually it should not matter. The design is such that if you know the actor's owner, you should be able to determine whether the actor is alive or not. The actor info is still saved in the GCS.
> 
> In this case should we preserve the current logic for handling node failure? Otherwise if the actor's owner and the actor resides on the same node, nobody will be reconstructing the specific actor when the node fails because its owner would die as well. We can remove the logic after this issue is resolved later. What do you think?

I don't think we should block this PR just to resolve that issue. The way to handle that particular issue is for the owner's owner to restart it. Once the owner is restarted, it should also restart the child actor.

If there are more questions about the actor supervision approach, can you please open a github issue for it? Thanks!

**Comment by zhijunfu** [comment]
> > > > I didn't have chance to look at most of the implementations yet. A quick question:
> > > > Do we assume the actor's owner to be alive in the node failure case? e.g. If the actor's owner resides on the same node with the actor. Does the actor owner saves the actor info that it created to some persistent place? e.g. GCS.
> > > 
> > > 
> > > Right now, yes, but eventually it should not matter. The design is such that if you know the actor's owner, you should be able to determine whether the actor is alive or not. The actor info is still saved in the GCS.
> > 
> > 
> > In this case should we preserve the current logic for handling node failure? Otherwise if the actor's owner and the actor resides on the same node, nobody will be reconstructing the specific actor when the node fails because its owner would die as well. We can remove the logic after this issue is resolved later. What do you think?
> 
> I don't think we should block this PR just to resolve that issue. The way to handle that particular issue is for the owner's owner to restart it. Once the owner is restarted, it should also restart the child actor.
> 
> If there are more questions about the actor supervision approach, can you please open a github issue for it? Thanks!

The problem is existing scenarios will be broken. It's okay if you will address it in your next PR sometime soon.

**Comment by ericl** [comment]
@zhijunfu do you have a test case for this scenario? Seems easy to reproduce in Python by killing processes. It would be helpful to add this to the Travis tests.

**Comment by zhijunfu** [comment]
> @zhijunfu do you have a test case for this scenario? Seems easy to reproduce in Python by killing processes. It would be helpful to add this to the Travis tests.

@ericl There's an existing test case [test_actor_reconstruction_on_node_failure](https://github.com/ray-project/ray/blob/master/python/ray/tests/test_actor.py#L2303), but it might not cover the scenario where the actor and its owner are on the same node though. Agreed it would be nice to have a specific case for this.

File: src/ray/core_worker/actor_manager.cc
Comment (zhijunfu) [reply src/ray/core_worker/actor_manager.cc]: Could you explain a bit more on when this would happen? Thanks!

**Comment by stephanie-wang** [comment]
> > @zhijunfu do you have a test case for this scenario? Seems easy to reproduce in Python by killing processes. It would be helpful to add this to the Travis tests.
> 
> @ericl There's an existing test case [test_actor_reconstruction_on_node_failure](https://github.com/ray-project/ray/blob/master/python/ray/tests/test_actor.py#L2303), but it might not cover the scenario where the actor and its owner are on the same node though. Agreed it would be nice to have a specific case for this.

Thanks, @zhijunfu, it looks like that test is failing on travis. I'll hold off on merging this for now. Planning on opening another PR soon that will pass through the address of the owner. Once we merge that in, we should be able to fix the failing test by checking if the owner has died.

**Comment by stephanie-wang** [comment]
Hmm actually looks like the test failures were due to a different issue (I think since #6088 was merged).

@zhijunfu, it'd be great if you could write a test case for the scenario you mentioned. I'm planning on merging this if travis is passing after the latest push.

**Comment by zhijunfu** [comment]
@stephanie-wang  Yep, will do 

> @zhijunfu, it'd be great if you could write a test case for the scenario you mentioned. I'm planning on merging this if travis is passing after the latest push.

**Comment by zhijunfu** [comment]
https://github.com/ray-project/ray/pull/6157

> @stephanie-wang Yep, will do
> 
> > @zhijunfu, it'd be great if you could write a test case for the scenario you mentioned. I'm planning on merging this if travis is passing after the latest push.

**Comment by stephanie-wang** [comment]
Closing for now to avoid breaking the raylet codepath while we work on direct tasks.

**Pull request closed without merging**
